### PR TITLE
Discrete transitions

### PR DESCRIPTION
**Pull request by mforets**
This is a start on handling discrete transitions for hybrid systems. Instead of storing the whole `(R(X ∩ G ∩ I⁻) ⊕ W) ∩ I⁺` as a LazySet, we see this as a function `X -> post_d(X)`. With this in mind we can easily dispatch on a given heuristics to compute this set (exactly if possible, or using template hulls, using a box approximation, combination of these, etc), without actually changing the hybrid loop code. 

This PR has some cases to handle unions of polyhedra by using the [distributive property](https://en.wikipedia.org/wiki/Algebra_of_sets) for the intersection of unions.

Other important methods are missing, e.g. template hull overapproximation a la Frehse & Ray (https://www.sciencedirect.com/science/article/pii/S1474667015371809), so we can combine this with support function based reachability algorithms (LGG09).. but i would leave that for another PR.

### CODE REVIEW

File: src/Hybrid/transitions.jl [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+reset_map(DT::DiscreteTransition) = ConstrainedAffineMap(R, W)
+guard(DT::DiscreteTransition) = DT.G
+source_invariant(DT::DiscreteTransition) = DT.I⁻
+target_invariant(DT::DiscreteTransition) = DT.I⁺
+
+# constructor when the reset map is the identity
+function DiscreteTransition(; guard::GT, source_invariant::IT⁻, target_invariant::IT⁺) where {GT, IT⁻, IT⁺}
+    n = dim(source_invariant)
+    return DiscreteTransition(Universe(n), ZeroSet(n), guard, source_invariant, target_invariant)
```
Comment (schillic) [inline src/Hybrid/transitions.jl:41]: I think the first argument (`R`) should be the identity map...

File: src/Hybrid/transitions.jl [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+# Cases where the reset map is the identity
+# ==========================================
+
+# ---------------------------
+# Intersection method: Exact
+# ---------------------------
+
+# all sets are polyhedral
+function _apply(tr::DiscreteTransition{<:Universe, <:ZeroSet, GT, IT⁻, IT⁺},
+                X::AbstractPolyhedron{N},
```
Comment (schillic) [inline src/Hybrid/transitions.jl:58]: It is probably more useful to dispatch on `AbstractPolytope` and then also return an `HPolytope`.

File: src/Hybrid/transitions.jl (hunk 1 above)
Comment (mforets) [reply src/Hybrid/transitions.jl:41]: hmm, true! we can use `MathematicalSystems.IdentityMap`

File: src/Hybrid/transitions.jl [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+                X::ConvexHullArray{N, PT},
+                ::Exact) where {N,
+                                PT<:AbstractPolyhedron{N},
+                                GT<:AbstractPolyhedron{N},
+                                IT⁻<:AbstractPolyhedron{N},
+                                IT⁺<:AbstractPolyhedron{N}}
+
+    vlist = [vertices_list(Xi) for Xi in array(X)]
+    vlist_ch = convex_hull(vcat(vlist...)) # TODO pass the backend as an option
+    Xhrep = tohrep(vlist_ch)
```
Comment (schillic) [inline src/Hybrid/transitions.jl:124]: I think this should be sent to `LazySets` as a first implementation of `constraints_list(X)`.

File: src/Hybrid/transitions.jl [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+                X::ConvexHullArray{N, PT},
+                ::BoxEnclosure) where {N,
+                                       PT<:AbstractPolyhedron{N},
+                                       GT<:AbstractPolyhedron{N},
+                                       VN,
+                                       IT⁻<:UnionSetArray{N, LinearConstraint{N, VN}},
+                                       IT⁺<:AbstractPolyhedron{N}}
+    Xbox = overapproximate(X, Hyperrectangle)
+    out = _apply(tr, Xbox, Exact())
+    return overapproximate(out, Hyperrectangle)
```
Comment (schillic) [inline src/Hybrid/transitions.jl:143]: Hm, I am not sure about the first box approximation. Why is that needed?

File: src/Hybrid/transitions.jl (hunk 2 above)
Comment (schillic) [reply src/Hybrid/transitions.jl:58]: Yes, but the set `X` is always bounded.

Comment (mforets) [reply src/Hybrid/transitions.jl:58]: oh, i see. indeed, i missed that argument.

File: src/Hybrid/transitions.jl (hunk 4 above)
Comment (mforets) [reply src/Hybrid/transitions.jl:143]: hmm i'm also not convinced about this choice. i think the most reasonable thing to do (at least as a fallback implementation) would be to evaluate the support functions of the lazy intersection array along the box directions (as in https://github.com/JuliaReach/ReachabilityAnalysis.jl/issues/172). . we can revisit this later.

Comment (mforets) [reply src/Hybrid/transitions.jl:143]: the benefit of using `Xbox = overapproximate(X, Hyperrectangle)` is that you don't have to make the possibly expensive `constraints_list(X)`.

Comment (schillic) [reply src/Hybrid/transitions.jl:143]: > the benefit of using Xbox = overapproximate(X, Hyperrectangle) is that you don't have to make the possibly expensive constraints_list(X).

True, but that applies to all other methods as well. With this initial overapproximation, the result is generally not the tightest box approximation. I find it rather arbitrary to do it here. It is certainly one possibility to do that but it may have unexpected results.

Comment (mforets) [reply src/Hybrid/transitions.jl:143]: ok. yes, `BoxEnclosure` should return the tightest box enclosure (if possible).

**Comment by mforets** [comment]
The last commit contains some missing methods to handle `UnionSetArray` (we can remove them again once we admit this type as a `<:LazySet)`. There are now to abstract interfaces, `AbstractIntersectionMethod` for methods to compute the intersection, and  `AbstractDisjointnessMethod` for methods to compute disjointness. This incorporates some changes that we discussed above regarding what is a reasonable box aproximation: in fact, you can expect that `SupportFunctionIntersection` is a struct that has a template field, say `::BoxDirections`, and we can just it it (as we can use  any other template).
